shopinvader/services/abstract_download.py

```diff
@@ -0,0 +1,113 @@
+# -*- coding: utf-8 -*-
+# Copyright 2020 ACSONE SA/NV (<http://acsone.eu>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+import mimetypes
+
+from odoo import _
+from odoo.addons.base_rest.components.service import (
+    skip_secure_response,
+    to_int,
+)
+from odoo.addons.component.core import AbstractComponent
+from odoo.exceptions import MissingError
+from odoo.http import content_disposition, request
+
+
+class AbstractDownload(AbstractComponent):
+    """
+    Class used to define behaviour to generate and download a document.
+    You only have to inherit this AbstractComponent and implement the function
+    _get_report_action(...).
+
+    Example for invoice service:
+    Class InvoiceService(Component):
+        _inherit = [
+            "base.shopinvader.service",
+            "abstract.shopinvader.download",
+        ]
+        _name = "shopinvader.invoice.service"
+
+        def _get_report_action(self, target, params):
+            return target.invoice_print()
+    """
+
+    _name = "abstract.shopinvader.download"
+
+    # This function should be overwritten
+    def _get_report_action(self, target, params=False):
+        """
+        Get the action/dict to generate the report
+        :param target: recordset
+        :param params: dict
+        :return: dict/action
+        """
+        raise NotImplementedError()
+
+    def _validator_download(self):
+        return {"id": {"type": "integer", "required": True, "coerce": to_int}}
+
+    @skip_secure_response
+    def download(self, _id, **params):
+        """
+        Get target file. This method is also callable by HTTP GET
+        """
+        target = self._get(_id)
+        headers, content = self._get_binary_content(target)
```

Comment:
@acsonefho you should mast params to all the methods defined into this abstract...